metrics.go

```diff
@@ -0,0 +1,63 @@
+/*
+ * Nuts go core
+ * Copyright (C) 2019 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package core
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/labstack/echo/v4"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const nutsMetricsPrefix = "nuts_"
+
+//NewMetricsEngine creates a new Engine for exposing prometheus metrics via http.
+//Metrics are exposed on /metrics, by default the GoCollector and ProcessCollector are enabled.
+func NewMetricsEngine() *Engine {
+	return &Engine{
+		Name:      "Metrics",
+		Configure: configure,
+		Routes: func(router EchoRouter) {
+			router.GET("/metrics", echo.WrapHandler(promhttp.Handler()))
+		},
+	}
+}
+
+var nutsRegistry prometheus.Registerer
+
+// Registerer returns a prometheus register with the engine name added as prefix. The nuts_ prefix has also been added.
+func Registerer(engine *Engine) prometheus.Registerer {
```

Comment:
I'm thinking, this approach (engine registers its own metrics in Metrics engine) v.s. the Engine exposing a RegisterMetrics callback like Configure/Routes should we consider it a 'core concern' like 'routes'.

Reply:
It's a thin layer indeed. Metrics is a bit like logging.

The most basic setup would have been to use the prometheus lib and not create a metrics engine at all. Then you could call `prometheus.NewX` from anywhere and set the metrics prefix yourself. My line of thought went from there but add a helper to create the correct prefix based on name and some preconfigured metrics.

The simplest solution would be to remove the helper, this would allow every piece of code to register metrics. This would remove some spaghetti of passing along the `registerer` through several layers of code.

Reply:
Nuts go core providing auxillery services (status, logging and now metrics) is fitting I think. Since engines are services that run inside that defined framework.

My though was whether this is the right way to register it since introduces another singleton. But we have that all over the place, so one more here wouldn't matter too much.w.

Can't think of a much better way right now. :heavy_check_mark: